File: leetcode/105-ConstructBinaryTreefromPreorderandInorderTraversal/constructBinaryTreefromPreorderandInorderTraversal.c
/*          
 * [Source] 
 *          
 * https://leetcode.com/problems/construct-binary-tree-from-preorder-and-inorder-traversal/#/description
 *          
 * [Problem Description]
 *          
 * Given preorder and inorder traversal of a tree, construct the binary tree.
 * Note:
 * You may assume that duplicates do not exist in the tree.
 *          
 * [Comments]
 *          
 *          
 *          
 * [Companies]
 *
 * Microsoft
 */          


/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
struct TreeNode* buildTreeHelper(int* preorder, int preStart, int preEnd, int* inorder, int inStart, int inEnd);

struct TreeNode* buildTree(int* preorder,
                           int preorderSize,
                           int* inorder,
                           int inorderSize)
{
  return buildTreeHelper(preorder, 0, preorderSize - 1, inorder, 0, inorderSize-1);
}

struct TreeNode* buildTreeHelper(int* preorder,
                                 int preStart,
                                 int preEnd,
                                 int* inorder,
                                 int inStart,
                                 int inEnd)
{
  if (preStart > preEnd || inStart > inEnd)
    return NULL;
  struct TreeNode* root = malloc(sizeof(struct TreeNode));
  root->val = preorder[preStart];
  root->left = root->right = NULL;
  int inIndex = 0; // Index of the current root in int* inorder
  int i;
  for(i = inStart; i <= inEnd; i++)
  {
    if (inorder[i] == root->val)
    {
      inIndex = i;
      break;
    }
  }
  root->left = buildTreeHelper(preorder, preStart+1, preStart + inIndex - inStart, inorder, inStart, inIndex-1);
  root->right = buildTreeHelper(preorder, preStart + inIndex - inStart + 1, preEnd, inorder, inIndex+1, inEnd);
  return root;
}

File: leetcode/105-ConstructBinaryTreefromPreorderandInorderTraversal/constructBinaryTreefromPreorderandInorderTraversal.java
/*          
 * [Source] 
 *          
 * https://leetcode.com/problems/construct-binary-tree-from-preorder-and-inorder-traversal/#/description
 *          
 * [Problem Description]
 *          
 * Given preorder and inorder traversal of a tree, construct the binary tree.
 * Note:
 * You may assume that duplicates do not exist in the tree.
 *          
 * [Comments]
 *          
 * This solution uses a hashtable for quickly look up the index of the root
 * from pre-order list in in-order list
 *          
 * [Companies]
 * 
 * Microsoft
 */          


/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

public class Solution {
  public TreeNode buildTree(int[] preorder, int[] inorder) {
    Map<Integer, Integer> inMap = new HashMap<Integer, Integer>();
    
    for(int i = 0; i < inorder.length; i++) {
      inMap.put(inorder[i], i);
    }

    TreeNode root = buildTree(preorder, 0, preorder.length - 1, inorder, 0, inorder.length - 1, inMap);
    return root;
  }

  public TreeNode buildTree(int[] preorder, int preStart, int preEnd, int[] inorder, int inStart, int inEnd, Map<Integer, Integer> inMap) {
    if(preStart > preEnd || inStart > inEnd) return null;
    
    TreeNode root = new TreeNode(preorder[preStart]);
    int inRoot = inMap.get(root.val);
    int numsLeft = inRoot - inStart;
    
    root.left = buildTree(preorder, preStart + 1, preStart + numsLeft, inorder, inStart, inRoot - 1, inMap);
    root.right = buildTree(preorder, preStart + numsLeft + 1, preEnd, inorder, inRoot + 1, inEnd, inMap);
    
    return root;
  }
}
